### DISCUSSION TITLE
Non Mono version?

### DISCUSSION THREAD

**Discussion by wbwqq** [lennart-finke/GodoTeX#8] [Ideas]
Would be great if this plugin could work on the non Mono version of Godot, especially since it's the most popular version. Would it be easy / lot of work to make it work ? (preferably for Godot 4).

I wouldn't mind contributing, but I have no knowledge in making Godot extensions.

**Comment by lennart-finke** [comment]
Hi and thank you kindly for your interest!

Indeed, making the plugin work without C# would be quite difficult. At the start of the project I tried to come up with ways to make it work without C#, but without success. I totally agree that this is not ideal, although you know, it works as is.

If you were set on contributing in this direction, which would be really admirable, it would basically entail writing a LaTeX renderer or in GDScript, which is a lot of work. I made the decision to depend on a simple MathJax-style renderer via a Nuget package and just implement a nice front-end to that. However, I am of course open to any ideas that might prove me wrong and enable us to run without Mono.

It would also be much appreciated if you got GodoTeX to run on Godot 4 Mono, which I have yet been unable to do.

**Reply by wbwqq** [reply]
So it's not possible to compile a C# extension to make the plugin available for all Godot versions, I see.

Writing a whole LaTeX renderer does seem like too much work. 

Have you heard of GDNative (new in Godot 4) ? With Godot 4's GDNative it seems possible to write C or C++ code more easily to make extensions. Maybe that could be a direction to explore? To be able to use a LaTeX renderer in C or C++ to make an extension.

https://docs.godotengine.org/de/stable/tutorials/scripting/gdnative/what_is_gdnative.html#doc-what-is-gdnative-third-party-bindings

**Reply by lennart-finke** [reply]
Yes, I have heard of GDNative and tried using it in the creation of GodoTeX before switching to C#. GDNative is not new in Godot 4, but its specifications have indeed changed from Godot 3.

It would be possible to integrate a LaTeX renderer through it and not depend on Mono, yes! I didn't find a suitable renderer back then, but if someone were to try again, that would be a promising approach. For a popular and well-written C++-based add-on, I would like to point to https://github.com/CoaguCo-Industries/GodotSteam

**Reply by wbwqq** [reply]
Sorry, I was mistaken. GDExtension is new in Godot 4 (not GDNative which as you said existed in Godot 3), it's a new version of GDNative.

https://godotengine.org/article/introducing-gd-extensions/

Do you think that could help?

So one issue you had is that you didn't find a renderer in C++ or C. So if we could find one, that would help I guess? I've also seen that it's possible to write GDNative extensions (so possibly also with GDExtension) from other languages, such as Python, Rust, and a few others.

**Reply by lennart-finke** [reply]
Exactly! Yes indeed, it a slight simplification from GDNative, which is helpful.

I did find some suitable LaTeX engines back in the day:

1) https://github.com/goldsborough/latexpp
2) https://github.com/NanoMichael/MicroTeX

MicroTeX sounds considerably friendlier in terms of dependencies, I do not recall what the problem was in the end.

Lastly, I would like to warn against using languages other than C++/GDScript, as that would be a guaranteed dependency & distribution hell.

**Comment by Megalomaniak** [comment]
You know, just because you need to use the mono version doesn't mean you have to use C#. As far as I'm aware anyways.
